src/NodeApi/Native/JSNativeApi.Interop.cs

```diff
@@ -1478,358 +2796,2267 @@ internal static bool TryGetExport(ref nint field, [CallerMemberName] string func
             return true;
         }
 
-        private static napi_status CallInterop<TResult>(
-            ref nint field,
-            napi_env env,
-            out TResult result,
-            [CallerMemberName] string functionName = "")
-            where TResult : unmanaged
+        internal static nint StringToHGlobalUtf8(string s)
         {
-            nint funcHandle = GetExport(ref field, functionName);
-            var funcDelegate = (delegate* unmanaged[Cdecl]<
-                napi_env, nint, napi_status>)funcHandle;
-            fixed (TResult* result_native = &result)
-            {
-                return funcDelegate(env, (nint)result_native);
-            }
+            if (s == null) return default;
+            byte[] bytes = Encoding.UTF8.GetBytes(s);
+            nint ptr = Marshal.AllocHGlobal(bytes.Length + 1);
+            Marshal.Copy(bytes, 0, ptr, bytes.Length);
+            Marshal.WriteByte(ptr, bytes.Length, 0);
+            return ptr;
         }
 
-        private static napi_status CallInterop<TResult1, TResult2>(
-            ref nint field,
-            napi_env env,
-            out TResult1 result1,
-            out TResult2 result2,
-            [CallerMemberName] string functionName = "")
-            where TResult1 : unmanaged
-            where TResult2 : unmanaged
+        private static nint GetFunctionPointerForDelegateAndRootIt<TDelegate>(TDelegate d)
+            where TDelegate : notnull
         {
-            nint funcHandle = GetExport(ref field, functionName);
-            var funcDelegate = (delegate* unmanaged[Cdecl]<
-                napi_env, nint, nint, napi_status>)funcHandle;
-            fixed (TResult1* result1_native = &result1)
-            fixed (TResult2* result2_native = &result2)
-            {
-                return funcDelegate(env, (nint)result1_native, (nint)result2_native);
-            }
+            GCHandle.Alloc(d); // To make sure that the delegate is not collected.
```

Comment:
I guess this is cleaner than what I did to prevent the "JS invoke" callback delegates from being GC'd, which was to save them in static fields. See `JSNativeApi.s_invokeJS*` - maybe those should use this approach also?

Reply:
I initially implemented static fields too following your approach, but then I asked ChatGPT about the GetFunctionPointerForDelegate usage and how to keep the delegate alive and it gave code that used the `GCHandle.Alloc`. It is just amazing what it can do if we can ask it right questions. :)

I will see how to improve the code for callbacks in the follow up PR.